homeassistant/components/volvo/config_flow.py

```diff
@@ -0,0 +1,176 @@
+            return self.async_show_form(
+                step_id="reauth_confirm",
+                description_placeholders={CONF_NAME: self._get_reauth_entry().title},
+            )
+        return await self.async_step_user()
+
+    async def async_step_api_key(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Handle the API key step."""
+        errors: dict[str, str] = {}
+
+        if user_input is not None:
+            web_session = aiohttp_client.async_get_clientsession(self.hass)
+            token = self._config_data[CONF_TOKEN][CONF_ACCESS_TOKEN]
+            auth = ConfigFlowVolvoAuth(web_session, token)
+            api = VolvoCarsApi(web_session, auth, "", user_input[CONF_API_KEY])
+
+            try:
+                self._vins = await api.async_get_vehicles()
+            except VolvoApiException:
+                _LOGGER.exception("Unable to retrieve vehicles")
+                errors["base"] = "cannot_load_vehicles"
+
+            if not errors:
+                self._config_data |= user_input
+
+                if len(self._vins) == 1:
+                    # If there is only one VIN, take that as value and
+                    # immediately create the entry. No need to show
+                    # additional step.
+                    self._config_data[CONF_VIN] = self._vins[0]
+                    return await self._async_create_or_update()
+
+                if self.source in (SOURCE_REAUTH, SOURCE_RECONFIGURE):
+                    # Don't let users change the VIN. The entry should be
+                    # recreated if they want to change the VIN.
+                    return await self._async_create_or_update()
+
+                return await self.async_step_vin()
+
+        if user_input is None:
+            if self.source == SOURCE_REAUTH:
+                user_input = self._config_data = dict(self._get_reauth_entry().data)
+            elif self.source == SOURCE_RECONFIGURE:
+                user_input = self._config_data = dict(
+                    self._get_reconfigure_entry().data
+                )
+            else:
+                user_input = {}
+
+        schema = vol.Schema(
+            {
+                vol.Required(
+                    CONF_API_KEY, default=user_input.get(CONF_API_KEY, "")
+                ): str,
+            },
+        )
+
+        return self.async_show_form(
+            step_id="api_key", data_schema=schema, errors=errors
+        )
+
+    async def async_step_vin(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Handle the VIN step."""
+        if user_input is not None:
+            self._config_data |= user_input
+            return await self._async_create_or_update()
+
+        schema = vol.Schema(
+            {
+                vol.Required(CONF_VIN): SelectSelector(
+                    SelectSelectorConfig(
+                        options=self._vins,
```

Comment:
Perhaps use `SelectOptionDict` so that we can present a nicer label with the car "name, model. etc." and not only a vin number.